lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,468 @@
+  using ExtType = llvm::df_iterator_default_set<BlockNode *>;
+  using UGT = llvm::GraphTraits<llvm::Undirected<BlockNode *>>;
+  using udf_iterator = llvm::df_iterator<BlockNode *, ExtType, false, UGT>;
+
+  BlockNode *EntryNode = Graph.getEntryNode();
+  auto Begin = udf_iterator::begin(EntryNode);
+  auto End = udf_iterator::end(EntryNode);
+
+  size_t DFSNum = 0;
+  for (; Begin != End; Begin++) {
+
+    // We assign the DFS numbering
+
+    BlockNode *CurrentNode = *Begin;
+
+    NodeDFSNumMap[CurrentNode] = DFSNum;
+    DFSNumNodeMap[DFSNum] = CurrentNode;
+    DFSNum++;
+
+    // Each time we reach a new node during the `DFS` visit, we search for the
+    // edge that brougth us here, and that we therefore need to insert into the
+    // spanning tree.
+    // Retrieve the last-but-one element on the `VisitStack`, which is the
+    // candidate as the source node of the edge we are interested into adding to
+    // the spanning tree.
+    auto VisitStackSize = Begin.getPathLength();
+
+    // When we have a single node on the `VisitStack`, we cannot assign an edge
+    // to the spanning tree
+    if (VisitStackSize < 2)
+      continue;
+
+    BlockNode *SourceNode = Begin.getPath(VisitStackSize - 2);
+
+    // Search for the edge which connects `Source` and `CurrentNode`
+    for (auto &[Neighbor, Index] :
+         llvm::children_edges<llvm::Undirected<BlockNode *>>(SourceNode)) {
+
+      // Check if the successor is equal to `CurrentNode`, and in case add it to
+      // the spanning tree
+      if (Neighbor == CurrentNode) {
+        SpanningTreeEdges.insert(Index);
+        break;
+      }
+    }
+  }
+
+  // Printout the spanning tree
+  revng_log(CycleEquivalenceAnalysisLogger, "");
+  revng_log(CycleEquivalenceAnalysisLogger,
+            "Spanning tree is composed by edges:");
+  for (auto &Index : SpanningTreeEdges) {
+    revng_log(CycleEquivalenceAnalysisLogger,
+              std::to_string(Index.getValue()) << "\n");
+  }
+}
+
+/// Helper function to compute the capping backedges needed values
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::computeCappingValues(BlockNode *BB,
+                                                                size_t &HI0,
+                                                                size_t &HI2) {
+
+  // 1) Compute the highest reached node from a backedge starting from the
+  //    current `Block`
+  HI0 = SizeTMaxValue;
+
+  // We are only interested in looking at backedges
+  for (EdgeDescriptor BackwardEdge : getBackwardEdges(BB)) {
+    BlockNode *Target = BackwardEdge.second;
+    HI0 = std::min(HI0, NodeDFSNumMap.at(Target));
+  }
+
+  // 2) Compute the highest reached node for any child of the current `Block`
+  size_t HI1 = SizeTMaxValue;
+  for (EdgeDescriptor ForwardEdge : getForwardEdges(BB)) {
+    BlockNode *Target = ForwardEdge.second;
+    HI1 = std::min(HI1, HighestReachedDFSNum.at(Target));
+  }
```

Comment:
I think the reasoning behind your comment is misguided. Indeed, `Target` are not the nodes reached by the backedges starting from `BB`, but are nodes reached by forward edges from BB. Those nodes are guaranteed to have been already processed (since are working in postorder), and therefore the corresponding entry in `HighestReachedDFSNum` for `Target` is populated (`HighestReachedDFSNum` is a "global" object shared between all the iterations).

Your comment was anyway spot-on, because I think that you were trying to say something like "what happens with a self-loop?". As elaborated above, this is not a bug here, but indeed I noticed that self-loops edge did not end up in any equivalence class, and I therefore fixed such behavior.

Reply:
Yeah you rephrased it better.

In any case this wasn't a bug. Closing this conversation.